Trac 61269: Add new filter and control for auto-redirect in 6.5.4

hellofromtonya opened the pull request:
* Adds a new `plugin_configuration_data_{$slug}` filter for plugins to register their URL.
* Adds redirect control. For 6.5.4, redirect will happen after plugin activation.

## Testing

[Instructions are found here in the Trac ticket](https://core.trac.wordpress.org/ticket/61269#comment:13).

## Sample code for plugins:

Plugins will need to add code to hook into / register their URL. Here's an example with WooCommerce where the `url` is the an absolute to its wizard:

```php
add_filter( 'plugin_configuration_data_woocommerce', static function() {
	return array(
		'url' => admin_url( 'admin.php?page=wc-admin' ),
	);
} );
``` 

Props @costdev @aaronjorbin 

Trac ticket: https://core.trac.wordpress.org/ticket/61269

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Add new filter and controls.

Props costdev.

costdev commented on src/js/_enqueues/wp/updates.js:
```diff
@@ -2447,7 +2456,7 @@
 
 	$( function() {
-		var $pluginFilter        = $( '#plugin-filter, #plugin-information-footer' ),
+		var $pluginFilter        = $( '.plugin-title .row-actions, #plugin-filter, #plugin-information-footer' ),
```
```suggestion
		var $pluginFilter        = $( '#plugin-filter, #plugin-information-footer' ),
```
This was some related test code that shouldn't be in the final patch. My mistake in including it in the `.diff`.

costdev commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
+	if ( is_array( $configuration_data ) ) {
+		$status['configurationData'] = array();
+
+		if ( isset( $configuration_data['url'] ) ) {
```
Should we add `wp_send_json_error()` calls here for the following cases?

- `$configuration_data` is not an array.
- `$configuration_data['url']` is set but isn't a string.
- `$configuration_data['url']` is set, is a string, but doesn't start with `admin_url()`?

hellofromtonya replied:
Hmm, debating ... While error handling is nice, is it necessary at this stage? Does it add value? Yes, it could for plugin devs, but maybe less useful for users?

hellofromtonya replied:
Also, it would add more strings to be translated, strings that likely may change during the 6.6 scope as we learn more.

costdev replied:
I guess it's really about whether invalid configuration data should result in a failed request, or be skipped and return a successful one. If we skip, an error message isn't needed. However, if we make the request fail, we need an error message.

So, if plugin configuration data is filtered with an invalid value, should that result in a failed request?

1. The plugin is activated, so the user action is successful.
2. It's the plugin author's action that will be unsuccessful, so it might be useful to inform them.
3. Despite it being the plugin author's action that fails, this also means the user is left without that next step.

I'm currently soft-leaning towards "it should fail", but with a message like "Plugin activated successfully, but the next step cannot be determined.", or something better worded.

hellofromtonya replied:
That's a good way to think it.

Right now, it skips it. Lean approach and less risky (less code) for the minor. Plus, it doesn't lock in the response or error messaging.

Then in the major with the onboarding framework, this area can be revisited to determine if it should fail and, if yes, then add a fail with error messaging.

What do you think?

costdev replied:
Alternatively, we just continue to display the 6.5.3 notice and the refresh should then show new menu items or some other change the plugin author has made.

costdev replied:
I'm happy to go with no failure or error message for the moment if we're confident that any issues will be caught during a plugin's development and not only fail depending on how some sites are configured.

hellofromtonya replied:
That's a good idea. Keep what's there for now.

So if the plugin registers invalid config data:

- The plugin should still activate.
- Auto-redirect will not happen.
- "Refresh now" user action and notice are then the fallback.

costdev commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
+		if ( isset( $configuration_data['url'] ) ) {
+			$configuration_url = trim( $configuration_data['url'] );
+
+			if ( str_starts_with( $configuration_url, admin_url() ) ) {
```
It may be worth changing `admin_url()` to `network_admin_url()`, which falls back to `admin_url()` anyway if it isn't a multisite [Ref](https://developer.wordpress.org/reference/functions/network_admin_url/). What do you think?

hellofromtonya replied:
Maybe? @JJJ what do you think?

afragen replied:
Not @JJJ but this makes sense.

JJJ replied:
I'm not confident this approach will work for plugins that could be activated under any of the site/network/user admin area contexts (_pretend for now I didn't just say per-user plugin activations_ 😅).

BuddyPress, for example, can be active on a single site, many individual sites, or the entire network, resulting in a different URL being passed into this filter based on where activation happened.

In a multisite installation:

- If I activate BuddyPress on a single site, using network_admin_url() would fail.
- If I activate BuddyPress on a network, using admin_url() would fail.

If we are going to ask plugins to pass in their own URL, why aren't we just trusting their values? 

If we are trying to avoid plugins doing unsafe redirects here, they can still do them a million other ways (and they could've always done them previously; any evidence of this as a problem?)

Does this URL need to be validated at all, and if so, would a home_url() check be sufficient (to make it context-agnostic) maybe using something like wp_http_validate_url()?

hellofromtonya replied:
What if ...

Instead of validating, it uses `sanitize_url()` for sanitizing "for .. redirect usage".

It then goes into `esc_url()` which cleans it with some validation (e.g. protocol).

This approach:
* is used in the same file in `wp_ajax_send_link_to_editor()` and `_wp_ajax_delete_comment_response()`.
* removes the `trim` and `str_starts_with( $configuration_url, admin_url() )` check to simplify the code.

How about this?

```php
if ( isset( $configuration_data['url'] ) && is_string( $configuration_data['url'] ) ) {
	$configuration_url = sanitize_url( $configuration_data['url'] );

	if ( $configuration_url ) {
		$status['configurationData']['url'] = $configuration_url;
	}
}
```

`sanitize_url()` will also return a `string`. Thus, the check before assignment skips if an empty string is returned.

hellofromtonya replied:
Is more validation needed such as ensuring it's in the admin?

Looking at the same file where URL is passed to the AJAX, there's no additional validation.

Looking at `wp_redirect()` and `wp_safe_redirect()`, there's no additional validation of filtered URLs.

Thinking no additional validation is needed.

hellofromtonya replied:
Applied the `sanitize_url()` approach in https://github.com/WordPress/wordpress-develop/pull/6606/commits/86cb5912f2510557d514851835f70d468732dacc.

Added a defensive guard to ensure the URL is a string (i.e. for better PHP 8 compat) as `sanitize_url()` requires a `string` https://github.com/WordPress/wordpress-develop/pull/6606/commits/3a4e54aafdaaa25eb876a650b68dd7809ff40658.

peterwilsoncc replied:
Should the redirect destination be passed through `wp_sanitize_redirect()` and `wp_validate_redirect()`? If it were a PHP redirect the practice would be to use `wp_safe_redirect()`.

As the code comes from a plugin, the developer can pretty much do anything but this seems like a case in which we can protect core from coding errors.

hellofromtonya committed:
Remove local testing code - whoopsie

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
+			$configuration_url = trim( $configuration_data['url'] );
+
+			if ( str_starts_with( $configuration_url, admin_url() ) ) {
+				$status['configurationData']['url'] = $configuration_url;
```
For security:

```suggestion
				$status['configurationData']['url'] = sanitize_url( $configuration_url );
```

`sanitize_url()` purpose: 
>Sanitizes a URL for database or redirect usage.

sybrew replied:
FYI: https://developer.wordpress.org/reference/functions/wp_sanitize_redirect/

hellofromtonya committed:
Use sanitize_url()

hellofromtonya committed:
URL is_string() defensive guard.

`sanitize_url()` expects a `string`. This defensive guard ensures a `string`
URL was passed.

Improves PHP 8 compatibility.

hellofromtonya commented:
# Test Report

Testing WooCommerce with this PR applied and sample code added to the plugin.

## Environment

- OS: macOS
- Web Server: nginx
- PHP: 8.3
- WordPress: 6.5.3 with this PR applied
- Browser: Firefox 126.0
- Theme: default
- Plugins: WooCommerce 8.9.1

## Instructions

Set Up:
1. Install but do not activate WooCommerce yet as code needs to be added to it.
2. Open `plugins/woocommerce/woocommerce.php` file.
3. Copy and paste the following the code after `defined( 'ABSPATH' ) || exit;`:

```php
add_filter( 'plugin_configuration_data_woocommerce', static function() {
	return array(
		'url' => admin_url( 'admin.php?page=wc-admin' ),
	);
} );
```

### Test in the Add Plugins UI:
1. Navigate to Plugins > Add New Plugin.
2. In the plugins search box, type `woocommerce`. The plugin cards will appear including the WooCommerce plugin.
3. In the WooCommerce card, press the "Activate" button.

Expected Results:
After activation, the site should redirect to Woo's onboarding wizard.

### Test in the More Details modal:
1. Deactivate WooCommerce first.
2. Navigate to Plugins > Add New Plugin.
3. In the plugins search box, type `woocommerce`. The plugin cards will appear including the WooCommerce plugin.
4. In the WooCommerce card, click on the "More Details" link below the "Activation" button.
5. Press the "Activate" button (may need to scroll down to the button).

Expected Results:
After activation, the site should redirect to Woo's onboarding wizard.

## Actual Results

### Test in the Add Plugins UI:
After activated, the onboarding wizard auto-redirects ✅ 

The "Refresh now" admin notice and button are visible during the time between "plugin activated" until the redirect triggers. This could confuse users. Marking it as not expected ❌.

![WooRedirect](https://github.com/WordPress/wordpress-develop/assets/7284611/def1fa8a-d99e-4bdf-a009-7049e65d432f)

### Test in the More Details modal:
After activated, the onboarding wizard auto-redirects ✅ 

The modal footer's "Refresh now" admin notice and button are visible during the time between "plugin activated" until the redirect triggers. This could confuse users. Marking it as not expected ❌.

![WooRedirect2](https://github.com/WordPress/wordpress-develop/assets/7284611/5ae7ee16-3105-4d04-9531-70c9729c69d2)

### Observations

"Refresh now" button and notice are visible ❌ (as noted in the tests results).

It's gone after navigating back (which is expected ✅ ).

hellofromtonya committed:
Prevent "Refresh now" rendering.

In the modal and parent UI, prevents the "Refresh now" notice
and button from rendering between the activated state and redirect.

hellofromtonya commented:
# Test Report

Testing WooCommerce with this PR applied (including the fix to not show the "Refresh now" https://github.com/WordPress/wordpress-develop/pull/6606/commits/dc013d3e1abbf3a826c02794d7e0c6b006d268ab) and sample code added to the plugin.

## Environment

- OS: macOS
- Web Server: nginx
- PHP: 8.3
- WordPress: 6.5.3 with this PR applied
- Browser: Firefox 126.0
- Theme: default
- Plugins: WooCommerce 8.9.1

## Instructions

Set Up:
1. Install but do not activate WooCommerce yet as code needs to be added to it.
2. Open `plugins/woocommerce/woocommerce.php` file.
3. Copy and paste the following the code after `defined( 'ABSPATH' ) || exit;`:

```php
add_filter( 'plugin_configuration_data_woocommerce', static function() {
	return array(
		'url' => admin_url( 'admin.php?page=wc-admin' ),
	);
} );
```

### Test in the Add Plugins UI:
1. Navigate to Plugins > Add New Plugin.
2. In the plugins search box, type `woocommerce`. The plugin cards will appear including the WooCommerce plugin.
3. In the WooCommerce card, press the "Activate" button.

Expected Results:
After activation:
* the "Refresh now" admin notice and button should not appear.
* the site should redirect to Woo's onboarding wizard.

### Test in the More Details modal:
1. Deactivate WooCommerce first.
2. Navigate to Plugins > Add New Plugin.
3. In the plugins search box, type `woocommerce`. The plugin cards will appear including the WooCommerce plugin.
4. In the WooCommerce card, click on the "More Details" link below the "Activation" button.
5. Press the "Activate" button (may need to scroll down to the button).

Expected Results:
After activation:
* at the bottom of the modal, the "Refresh now" admin notice and button should not appear.
* the site should redirect to Woo's onboarding wizard.

## Actual Results

### Test in the Add Plugins UI:
After activated:
* the "Refresh now" admin notice and button do not appear ✅ 
* the onboarding wizard auto-redirects ✅ 

### Test in the More Details modal:
After activated:
* the "Refresh now" footer area in the modal does not appear ✅ 
* the onboarding wizard auto-redirects ✅ 

### Observations
Works as expected  ✅

hellofromtonya commented:
This PR is ready for review and testing. Testing instructions are found in the description and ticket.

afragen commented on src/js/_enqueues/wp/updates.js:
```diff
@@ -1152,19 +1153,21 @@
+				}
 			}
-		} else {
+		} else if ( 'undefined' === typeof response.configurationData.url ) {
```
Not certain I understand that else conditional is same as initial conditional.

costdev replied:
This conditional is an `else` to the `isInModal` conditional on line 1144

jeherve commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
+	if ( is_array( $configuration_data ) ) {
+		$status['configurationData'] = array();
+
+		if ( isset( $configuration_data['url'] ) && is_string( $configuration_data['url'] ) ) {
```
What would you think about switching to an `empty` check instead of `isset` here, so we don't even bother doing anything if a URL is not provided?

```suggestion
		if ( ! empty( $configuration_data['url'] ) && is_string( $configuration_data['url'] ) ) {
```

jeherve reviewed (commented):
This tested well for me. I only have a minor comment on the filter's checks.

 # Test Report

I tested with the current version of this branch applied, with 2 different plugins:

- One where I added the sample code to the plugin after installing it, before activating it.
- One where I did not add the filter, to test that the current behavior remains for plugins that will not make any changes.

## Environment

    OS: Android 
    Web Server: nginx
    PHP: 8.2
    WordPress: 6.5.3 with this PR applied
    Browser: Firefox for Android
    Theme: Björk
    Plugins: Color Posts

### Test in the Add Plugins UI:

This tested well, I was immediately redirected upon activation.

### Test in the More Details modal:

This tested well as well. I should note that the layout isn't perfect, we miss a bit of padding below the "Activate" button. That's not something that this PR introduces though, so not something to fix here.

![Screenshot_20240524-111519~2](https://github.com/WordPress/wordpress-develop/assets/426388/41923e8e-1a5d-43a7-820d-8b6090fb1fc3)

jeherve commented:
I wonder if we should change `url` in the filter parameters into something more specific?

I could imagine this filter being extended in the future with additional parameters, and maybe we would end up with a `settings_url` and a `onboarding_url` for example:

- `settings_url` being used for URLs such as the "Settings" links plugin authors will often add to the plugin list table with the `plugin_action_links_{$plugin_file}` filter
- `onboarding_url` being a unique URL to start onboarding when the plugin is first activated, like Yoast SEO's `wp-admin/admin.php?page=wpseo_installation_successful_free`.


Edit: Ah, I see now this is also being discussed [on the ticket](https://core.trac.wordpress.org/ticket/61269#comment:27). Consider this a +1 then :)

_**Follow-up**_: After discussion in the ticket, [consensus is to use the `url` parameter in the minor](https://core.trac.wordpress.org/ticket/61269#comment:40).

hellofromtonya committed:
Replace isset with empty to skip if empty

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

adrianduffell commented:
Thanks for working on this, @hellofromtonya -- it is testing well for me in the WooCommerce plugin so far.

Should WordPress keep track of whether the each plugin has completed onboarding, or should this logic be offloaded to the individual plugins? 🤔 

For WooCommerce, there is already logic in the `admin.php?page=wc-admin` URL keeping track of this. The page will determine whether configuration is incomplete and therefore display the onboarding over the home screen. This is necessary to prevent the onboarding unnecessarily showing again on "re-activation" (i.e. if the user performs the sequence: install > activate > onboarding > de-activate > activate). However it might be worth having WordPress know the completion state so this logic can be built into WordPress directly. If so, there could perhaps be an additional `completed` key on the array for plugins to declare whether configuration has been completed:

```
add_filter( 'plugin_configuration_data_woocommerce', static function() {
	return array(
		'url' => admin_url( 'admin.php?page=wc-admin' ),
		'completed' => some_function_returning_true_or_false(),
	);
} );
```

hellofromtonya commented:
Thanks for testing @adrianduffell.

>Should WordPress keep track of whether the each plugin has completed onboarding, or should this logic be offloaded to the individual plugins?

That's an interesting question. If/when onboarding is integrated into the workflow, that might well be a state Core will need to track, i.e. for bringing the experience back for the next plugin in the chain and/or altering users of the current state.

I'm thinking it's too early though to introduce it in this scope of work for the minor. 

Why? It needs discussion and consideration for how it and onboarding in general fits into the user + plug chain workflow. IMO it's better served in the [onboarding framework Trac ticket](https://core.trac.wordpress.org/ticket/61040).

Would you mind sharing that idea in  that that ticket please?

hellofromtonya commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,39 @@ function wp_ajax_activate_plugin() {
 	}
 
+	/**
+	 * Filters the configuration data for the plugin.
+	 *
+	 * The dynamic portion of the hook name refers to the plugin slug.
+	 *
+	 * How WordPress uses this data may change in the future.
+	 *
+	 * Currently, the user will be redirected to the configuration URL
+	 * after activating the plugin.
```
Let's discuss and iterate on improving the filter's description per @costdev [suggestion of focusing it on ""What am I filtering?"](https://core.trac.wordpress.org/ticket/61269#comment:39) rather than how Core will use the configuration data. 

Note: The filter is not driving the behavior, which is controlled by Core. Rather, it's the entry point where a plugin registers their configuration data with the install > activate > onboard workflows.

What's being filtered?

How about this?

```suggestion
	 * Filters the configuration data for the add plugins workflow.
	 *
	 * The dynamic portion of the hook name refers to the plugin slug.
```

@costdev what is a better way to describe " install > activate workflow"?

afragen replied:
Maybe “initial plugin activation workflow”?

hellofromtonya replied:
Adjusted using @afragen suggestion. Is this better and more clear?

hellofromtonya replied:
Thinking more about this .... "activation workflow" might lock into only being for activation. This could be limiting for when the next step is introduced, though one could argue, this is adding the entry point into the next step.

What if it's changed to the "add plugins workflow":

>Filters the configuration data for the add plugins workflow.

This wording aligns well to the Add Plugins UI where the cards and modal are. Then when the 3rd step is introduced in a later release, it could be expanded if needed.

@afragen @costdev and everyone, what do you think?

afragen replied:
Honestly I think any follow up steps should be handled by the plugin and not core. 

My reasoning is that there can be too many possible next steps. 

Core should handle the most common process and that's the initiation of the onboarding the plugin should be able to handle any subsequent processes.

costdev replied:
Note that this exact filter is maybe/probably/likely to be used to add a similar link in a plugin's action links on the `Plugins > Installed plugins` screen.

We can always update a filter description later, but something like this may be more fitting for the longer term:

> Filters data for a plugin's configuration workflow.

That way, it's not about the workflows that lead to this data, but rather the workflows that this data leads to.

hellofromtonya replied:
Aha, yes, that's better and on point @costdev.

Should it use the same language to include onboarding | set up | configuration?

>Filters data for a plugin's onboarding | set up | configuration workflow.

costdev replied:
Hmm, so technically, "set up" and "configuration" are usually the same, it's just that "set up" is like a "first-time configuration". Even if "configuration" points to more settings and "set up" is just the core configuration, using "configuration" should still work.

I'm mindful that if we add "onboarding" to the main subject of the docblock, then changing it later may be problematic.

We could do something like this maybe?
> Filters the data for the plugin's configuration workflow.
>
> In WordPress 6.5, this can be used to supply data for the plugin's onboarding workflow too.
> See https://core.trac.wordpress.org/ticket/61040 to follow future development on this.

What do you think?

costdev replied:
Thinking more on this, even if we introduce an onboarding framework later, this filter can be used to supply any URL, so maybe we don't need to detail different use cases here (beyond "configuration"), and just support the provision of a URL to be determined by the plugin author.

hellofromtonya replied:
>Hmm, so technically, "set up" and "configuration" are usually the same, it's just that "set up" is like a "first-time configuration". Even if "configuration" points to more settings and "set up" is just the core configuration, using "configuration" should still work.

The first time I read the onboarding ticket I thought - huh, onboarding is an experience for configuring. I've been using the 3 separate terms for consistency. But to me, they are all forms of configuration.

Thus, I agree with you. It's configuration. Even the "onboarding framework" ticket is about presenting users with the path towards "configuration".

 Now back to this filter's docblock.
 
I'm not sure this part adds enough value to be included:

>In WordPress 6.5, this can be used to supply data for the plugin's onboarding workflow too.
See https://core.trac.wordpress.org/ticket/61040 to follow future development on this.

What if that ticket gets closed such as not enough users want it or a different direction surfaces?

I'm thinking [what you proposed here works well](https://github.com/WordPress/wordpress-develop/pull/6606#discussion_r1617860270).

>Filters the data for the plugin's configuration workflow.

I'll push that now.

hellofromtonya committed:
Update filter docblock to what it filters

peterwilsoncc commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,32 @@ function wp_ajax_activate_plugin() {
+	 *                        Default empty string for no configuration URL.
+	 * }
+	 */
+	$configuration_data = apply_filters( 'plugin_configuration_data_' . $status['slug'], array( 'url' => '' ) );
```
```suggestion
	$configuration_data = apply_filters( "wp_plugin_configuration_data_{$status['slug']}", array( 'url' => '' ) );
```

* Prefixes the filter with WP because it's a WordPress filter
* CS fix per [coding standards for dynamic hooks](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#interpolation-for-naming-dynamic-hooks).

hellofromtonya replied:
Good catches! Thanks @peterwilsoncc. Suggestions done in https://github.com/WordPress/wordpress-develop/pull/6606/commits/d2749cd20858e7538fd27313d00cdb01d0f68e58.

peterwilsoncc commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4588,6 +4588,32 @@ function wp_ajax_activate_plugin() {
 	}
 
+	/**
+	 * Filters the data for the plugin's configuration workflow.
```
```suggestion
	 * Filters the data for the plugin's configuration workflow.
	 *
	 * The dynamic portion of the hook refers to the plugin's slug provided by the
	 * AJAX activation request. For the Hello Dolly plugin, this would be `hello-dolly`.
```

hellofromtonya replied:
Nice. I like that. Thanks @peterwilsoncc

peterwilsoncc reviewed (commented):
Docblocks, coding standards and naming things.

hellofromtonya committed:
Improve the filter description.

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

hellofromtonya committed:
Prefix filter. Combine dynamic hook var.

* Prefixes the filter with WP because it's a WordPress filter
* CS fix per coding standards for dynamic hooks.

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

hellofromtonya commented:
Closing in favor [Trac 61319](https://core.trac.wordpress.org/ticket/61319) and its patch #6670.

Thank you everyone for contributing to this effort 🥇  This work can continue in https://core.trac.wordpress.org/ticket/61040 as part of a major release.
